File: app/src/main/java/com/softaai/designpatterns/java/structural/flyweight/ShapeFactory.java
package com.softaai.designpatterns.java.structural.flyweight;

import java.util.HashMap;

/**
 * Created by Amol Pawar on 04-04-2019.
 * softAai Apps
 */
public class ShapeFactory {
    private static final HashMap<String, Shape> circleMap = new HashMap<>();
    public static Shape getCircle(String color){
        Circle circle = (Circle) circleMap.get(color);
        if(circle == null){
            circle = new Circle(color);
            circleMap.put(color, circle);
            System.out.println("Creating circle of color : " + color);
        }
        return circle;
    }
}

File: app/src/main/java/com/softaai/designpatterns/kotlin/structural/flyweight/ShapeFactory.kt
package com.softaai.designpatterns.kotlin.structural.flyweight


/**
 * Created by Amol Pawar on 04-04-2019.
 * softAai Apps
 */
object ShapeFactory {

    private val circleMap : HashMap<String, Shape> = HashMap()


    fun getCircle(color : String) :Shape{
        var circle = circleMap.get(color)
        if(circle == null){
            circle = Circle(color)
            circleMap.put(color, circle)
            println("Creating circle of color : " + color);
        }
        return circle
    }
}